include/micm/util/cuda_sparse_matrix.hpp

```diff
@@ -0,0 +1,91 @@
+#include <type_traits>
+#include <micm/util/sparse_matrix.hpp>
+#include <micm/util/cuda_vector_matrix.cuh>
+#include <micm/util/cuda_param.hpp>
+
+namespace micm {
+  template<class T, class OrderingPolicy>
+  class CudaSparseMatrix : public SparseMatrix<T, OrderingPolicy>
+  {
+private:
+    CudaVectorMatrixParam param_;
+
+public:
+    CudaSparseMatrix() = default;
+
+    CudaSparseMatrix(const SparseMatrixBuilder<T, OrderingPolicy>& builder) requires(std::is_same_v<T, double>)
+      : SparseMatrix<T, OrderingPolicy>(builder)
+    {
+      micm::cuda::MallocVector(param_, this->data_.size());
+    }
+    CudaSparseMatrix(const SparseMatrixBuilder<T, OrderingPolicy>& builder)
+      : SparseMatrix<T, OrderingPolicy>(builder)
+    {}
+
+    CudaSparseMatrix<T, OrderingPolicy>& operator=(const SparseMatrixBuilder<T, OrderingPolicy>& builder) requires(std::is_same_v<T, double>)
+    {
+      SparseMatrix<T, OrderingPolicy>::operator = (builder);
+      micm::cuda::MallocVector(param_, this->data_.size());
+      return *this;
+    }
+
+    CudaSparseMatrix<T, OrderingPolicy>& operator=(const SparseMatrixBuilder<T, OrderingPolicy>& builder)
+    {
+      SparseMatrix<T, OrderingPolicy>::operator = (builder);
+      return *this;
+    }
+
+    CudaSparseMatrix(const CudaSparseMatrix& other) requires(std::is_same_v<T, double>)
+      : SparseMatrix<T, OrderingPolicy>(other)
+    {
+      micm::cuda::MallocVector(param_, this->data_.size());
+      micm::cuda::CopyToDeviceFromDevice(param_, other.param_);
+    }
+
+    CudaSparseMatrix(const CudaSparseMatrix& other)
+      : SparseMatrix<T, OrderingPolicy>(other)
+    {}
+
+    CudaSparseMatrix(CudaSparseMatrix&& other) noexcept
+      : SparseMatrix<T, OrderingPolicy>(SparseMatrix<T, OrderingPolicy>::create(other.number_of_blocks_))
+    {
+      this->data_ = std::move(other.data_);
+      this->param_ = std::move(other.param_);
+    }
+
+    CudaSparseMatrix& operator=(const CudaSparseMatrix& other)
+    {
+      return *this = CudaSparseMatrix(other);
+    }
+
+    CudaSparseMatrix& operator=(CudaSparseMatrix&& other) noexcept
+    {
+      std::swap(this->data_, other.data_);
+      std::swap(this->param_, other.param_);
+      std::swap(this->number_of_blocks_, other.number_of_blocks_);
+      std::swap(this->row_ids_, other.row_ids_);
+      std::swap(this->row_start_, other.row_start_);
```

Comment:
what are these three variables for? should them be part of the `param_` struct and used on GPU later? If so, I guess we should use another Param struct rather than `CudaVectorMatrixParam`?

Reply:
So I didn't create another struct because I didn't see where these were being used on the GPU but if they are needed on the device, I should make a new struct for matrix params.

Reply:
@mattldawson can you provide an example where these three variables are used on CPU? I wonder maybe I miss a CUDA implementation somewhere in this case.

Reply:
Aha, thanks. I actually wonder how the index is mapped between the sparse matrix and vector in your yesterday's example and now I see these variables do the trick. Thank you for the clarification.